bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/ScanAndCompareGarbageCollector.java

```diff
@@ -138,6 +139,38 @@ public void gc(GarbageCleaner garbageCleaner) {
         }
     }
 
+    /**
+     * Cleans ledger safely by verifying ledger-metadata is deleted from the zk else it skips ledger-gc if ledger-metada
+     * exists in zk.
+     * 
+     * @param garbageCleaner
+     * @param ledgerId
+     * @param ledgerManager
+     * @param semaphore
+     * @throws InterruptedException
+     */
+    private void gcLedgerSafely(GarbageCleaner garbageCleaner, long ledgerId, LedgerManager ledgerManager,
+            Semaphore semaphore) throws InterruptedException {
+        semaphore.acquire();
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicBoolean ledgerDeleted = new AtomicBoolean(false);
+        ledgerManager.existsLedgerMetadata(ledgerId, (rc, exists) -> {
+            if (rc == BKException.Code.NoSuchLedgerExistsException) {
+                ledgerDeleted.set(true);
+            } else if (rc == BKException.Code.OK) {
+                LOG.warn("Can't delete ledger {} with metadata exists in zk ", ledgerId);
+            } else {
+                LOG.warn("Fail to check {} exists in zk {}", ledgerId, BKException.getMessage(rc));
+            }
+            latch.countDown();
+            semaphore.release();
```

Comment:
I don't think we need a semaphore at all.  Except during unit tests the gc method is called by `scheduleAtFixedRate`, which guarantees that later calls to gc will not happen until the previous ones finish.

if garbageCleaner.clean was inside the callback then the semaphore would make since, but with the CountDownLatch this is a blocking call, so it really doesn't.

Reply:
yes, i missed to remove it after adding latch. will fix it.